src/eigsolve/lanczos.jl

```diff
@@ -150,3 +150,139 @@ function eigsolve(A, x₀, howmany::Int, which::Selector, alg::Lanczos;
+            # Note: Fast eigen solver for block tridiagonal matrices is not implemented yet.
+            TDB = view(fact.TDB, 1:K, 1:K)
+            D, U = eigen(Hermitian(TDB))
+            by, rev = eigsort(which)
+            p = sortperm(D; by=by, rev=rev)
+            D, U = permuteeig!(D, U, p)
+
+            # detect convergence by computing the residuals
+            bs_r = fact.r_size   # the block size of the residual (decreases as the iteration goes)
+            r = residual(fact)
+            UU = U[(end - bs_r + 1):end, :]  # the last bs_r rows of U, used to compute the residuals
+            normresiduals = diag(UU' * block_inner(r, r) * UU)
+            normresiduals = sqrt.(real.(normresiduals))
+            converged = count(nr -> nr <= tol, normresiduals)
+            if converged >= howmany || β <= tol  # successfully find enough eigenvalues
+                break
+            elseif verbosity >= EACHITERATION_LEVEL
+                @info "BlockLanczos eigsolve in iteration $numiter: $converged values converged, normres = $(normres2string(normresiduals))"
+            end
+        end
+
+        if K < krylovdim
+            expand!(iter, fact; verbosity=verbosity)
+            numops += 1
+        else # shrink and restart
+            numiter >= maxiter && break
+            bsn = max(div(3 * krylovdim + 2 * converged, 5) ÷ bs, 1) # Divide basis into blocks with the same size
+            keep = bs * bsn
+            H = zeros(S, (bsn + 1) * bs, bsn * bs)
+            # The last bs rows of U contribute to calculate errors of Ritz values.
+            @inbounds for j in 1:keep
+                H[j, j] = D[j]
+                H[(bsn * bs + 1):end, j] = U[(K - bs + 1):K, j]
+            end
+            # Turn diagonal matrix D into a block tridiagonal matrix, and make sure 
+            # the residual of krylov subspace keeps the form of [0,..,0,R]
+            @inbounds for j in keep:-1:1
+                h, ν = householder(H, j + bs, 1:j, j)
+                H[j + bs, j] = ν
+                H[j + bs, 1:(j - 1)] .= zero(eltype(H))
+                lmul!(h, H)
+                rmul!(view(H, 1:(j + bs - 1), :), h')
+                rmul!(U, h')
+            end
+            # transform the basis and update the residual and update the TDB.
+            TDB .= S(0)
+            TDB[1:keep, 1:keep] .= H[1:keep, 1:keep]
+            B = basis(fact)
+            basistransform!(B, view(U, :, 1:keep))
+
+            r_new = OrthonormalBasis(fact.r.vec[1:bs_r])
+            view_U = view(U, (K - bs_r + 1):K, (keep - bs_r + 1):keep)
+            basistransform!(r_new, view_U)
+            fact.r.vec[1:bs_r] = r_new[1:bs_r]
+
+            while length(fact) > keep
+                pop!(fact.V)
+                fact.total_size -= 1
+            end
+            numiter += 1
+        end
+    end
+
+    howmany_actual = howmany
+    if converged > howmany
+        howmany_actual = converged
+    elseif length(D) < howmany
```

Comment:
Can this ever happen? Isn't `krylovdim` required to be larger than `howmany`?

Reply:
Yes, I also questioned my own implementation. But indeed, if the Krylov process reaches an invariant subspace of dimension smaller than `howmany`, it prints a warning at that place, but still relies on the rest of the code to extract the eigenvalues, so this needs to happen here.